File: src/lib/GemInterface.py
import os
import asyncio
from dotenv import load_dotenv
import requests
from requests.adapters import HTTPAdapter
from urllib3.util.retry import Retry
from bs4 import BeautifulSoup
from typing import Any, Optional, AsyncIterator
import json

from ollama import chat, AsyncClient, web_search
from ollama import ChatResponse

class AiInterface:
    """
    AI Interface using Ollama for local LLM inference with streaming support.
    Keeps the original synchronous web scraper (requests + BeautifulSoup) but improves it
    by adding a requests.Session with browser-like headers and a retry strategy to reduce
    403/429/5xx failures. Everything else remains async-friendly by running blocking
    operations in a threadpool.

    Usage:
      ai = AiInterface()
      result = asyncio.run(ai.Archie("When is fall break?"))
    """

    def __init__(
        self,
        debug: bool = False,
        scraper_max_retries: int = 3,
        scraper_backoff_factor: float = 1.0,
        scraper_timeout: int = 15,
    ):
        # Load the variables from the .env file into the environment
        load_dotenv()

        # Retrieve the model name from environment (defaults to llama2 if not set)
        self.model = os.getenv("MODEL", "llama2")

        # Debug flag
        self.debug = debug

        # Scraper configuration
        self.scraper_timeout = scraper_timeout

        # Create a requests.Session with headers and retry strategy
        self.session: requests.Session = requests.Session()
        # Browser-like default headers to reduce likelihood of 403 from simple bot checks
        self.session.headers.update({
            "User-Agent": (
                "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                "AppleWebKit/537.36 (KHTML, like Gecko) "
                "Chrome/120.0.0.0 Safari/537.36"
            ),
            "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
            "Accept-Language": "en-US,en;q=0.9",
        })

        # Configure retries for transient errors and common rate-limit statuses.
        retry_strategy = Retry(
            total=scraper_max_retries,
            status_forcelist=[429, 500, 502, 503, 504],
            allowed_methods=["HEAD", "GET", "OPTIONS"],
            backoff_factor=scraper_backoff_factor,
            raise_on_status=False,
        )
        adapter = HTTPAdapter(max_retries=retry_strategy)
        self.session.mount("https://", adapter)
        self.session.mount("http://", adapter)

    def _log(self, *args):
        if self.debug:
            print("[AiInterface DEBUG]", *args)

    def generate_text(self, prompt: str, system_prompt: str = "") -> str:
        """
        Synchronous method: generate text using Ollama client.
        This method is kept for backwards compatibility.
        """
        try:
            messages = []
            if system_prompt:
                messages.append({
                    'role': 'system',
                    'content': system_prompt,
                })
            messages.append({
                'role': 'user',
                'content': prompt,
            })
            
            response: ChatResponse = chat(model=self.model, messages=messages)
            return response['message']['content']
        except Exception as e:
            self._log(f"Error during text generation: {e}")
            return "An error occurred during text generation"

    def scrape_website(self, url: str, timeout: Optional[int] = None) -> str:
        """
        Improved synchronous web scraper that:
        - uses a persistent requests.Session with browser-like headers
        - has a Retry strategy for transient status codes (429, 5xx)
        - keeps the interface synchronous (requests + BeautifulSoup) as requested
        """
        try:
            to = timeout if timeout is not None else self.scraper_timeout
            self._log(f"Scraping {url} with timeout={to}")
            response = self.session.get(url, timeout=to, allow_redirects=True)
            # If raise_for_status raises, we catch below and return a helpful message
            try:
                response.raise_for_status()
            except requests.HTTPError as http_err:
                # Provide helpful debug string but still try to return body if available
                self._log(f"HTTP error for {url}: {http_err} (status {response.status_code})")
                # If server returned some HTML (e.g., Cloudflare block), parse and return its text
                # Otherwise, return the error string
                if response.text:
                    soup = BeautifulSoup(response.text, "html.parser")
                    return soup.get_text()
                return f"HTTP error when scraping {url}: {http_err}"

            soup = BeautifulSoup(response.text, 'html.parser')
            return soup.get_text()
        except requests.RequestException as e:
            self._log(f"RequestException when scraping {url}: {e}")
            return "An error occurred while scraping the website"
        except Exception as e:
            self._log(f"Unexpected error when scraping {url}: {e}")
            return "An unexpected error occurred while scraping the website"

    def web_search_ollama(self, query: str, max_results: int = 5) -> str:
        """
        Perform a web search using Ollama's built-in web search tool.
        This is used when scraped data doesn't contain the needed information.
        Note: Requires OLLAMA_API_KEY environment variable to be set.
        """
        try:
            self._log(f"Performing web search via Ollama for: {query}")
            response = web_search(query, max_results=max_results)
            
            if not response.results:
                return "No web search results found."
            
            formatted_results = []
            for i, result in enumerate(response.results, 1):
                title = result.title or 'No title'
                content = result.content or 'No description'
                url = result.url or 'No URL'
                formatted_results.append(f"{i}. {title}\n   {content}\n   Source: {url}")
            
            return "\n\n".join(formatted_results)
        except ValueError as e:
            # Ollama raises ValueError if OLLAMA_API_KEY is not set
            self._log(f"Web search requires OLLAMA_API_KEY: {e}")
            return "Web search unavailable: OLLAMA_API_KEY not configured"
        except Exception as e:
            self._log(f"Error during web search: {e}")
            return "An error occurred during web search"


    async def generate_text_async(self, prompt: str, system_prompt: str = "") -> str:
        """
        Async wrapper around Ollama chat for non-streaming responses.
        Runs the operation in a threadpool to avoid blocking the event loop.
        """
        def _generate_text(prompt: str, system_prompt: str = "") -> str:
            messages = []
            if system_prompt:
                messages.append({
                    'role': 'system',
                    'content': system_prompt,
                })
            messages.append({
                'role': 'user',
                'content': prompt,
            })
            
            try:
                response: ChatResponse = chat(model=self.model, messages=messages)
                return response['message']['content']
            except Exception as e:
                self._log(f"Error during text generation: {e}")
                return "An error occurred during text generation"
        
        loop = asyncio.get_running_loop()
        return await loop.run_in_executor(None, _generate_text, prompt, system_prompt)
    
    async def generate_text_streaming(self, prompt: str, system_prompt: str = "") -> AsyncIterator[str]:
        """
        Async streaming generator that yields tokens as they are generated by Ollama.
        This allows for real-time display of the AI's thinking process.
        
        Usage:
            async for token in ai.generate_text_streaming(prompt, system):
                print(token, end='', flush=True)
        """
        messages = []
        if system_prompt:
            messages.append({
                'role': 'system',
                'content': system_prompt,
            })
        messages.append({
            'role': 'user',
            'content': prompt,
        })
        
        try:
            # Create a new AsyncClient for each streaming request to avoid event loop conflicts
            async_client = AsyncClient()
            stream = await async_client.chat(
                model=self.model,
                messages=messages,
                stream=True
            )
            
            async for chunk in stream:
                if 'message' in chunk and 'content' in chunk['message']:
                    yield chunk['message']['content']
        except Exception as e:
            with open("error.txt", "a", encoding="utf-8") as f:
                f.write(f"Error during streaming: {e}\n")
            self._log(f"Error during streaming: {e}")
            yield "An error occurred during streaming"

    async def _run_in_executor(self, func: Any, *args, **kwargs):
        """
        Helper to run any synchronous function in the default threadpool and return its result.
        """
        loop = asyncio.get_running_loop()
        return await loop.run_in_executor(None, lambda: func(*args, **kwargs))
    
    async def Archie(self, query: str, conversation_history: list = None) -> str:
        """
        Main async entry point for the Archie AI assistant.
        Uses scraped data from JSON file to provide context for answering queries.
        Falls back to web search if needed.
        """
        with open("data/scrape_results.json", "r", encoding="utf-8") as f:
            results = json.load(f)
        
        # Build context with conversation history
        history_context = ""
        if conversation_history:
            history_context = "\n\nConversation History:\n"
            for msg in conversation_history[-5:]:  # Last 5 messages for context
                role = msg.get("role", "user")
                content = msg.get("content", "")
                history_context += f"{role.upper()}: {content}\n"
        
        # First attempt with scraped data
        system_prompt = f"""You are ArchieAI, an AI assistant for Arcadia University. You are here to help students, faculty, and staff with any questions they may have about the university.

You are made by students for a final project. You must be factual and accurate based on the information provided. It is ok to say "I don't know" if you are unsure.

Respond based on your knowledge up to 2025.

Use the following content to better answer the query:
{json.dumps(results, indent=2)}
{history_context}"""

        # Get initial response
        initial_response = await self.generate_text_async(query, system_prompt=system_prompt)
        
        # Check if we need to do a web search (if response indicates insufficient info)
        if any(phrase in initial_response.lower() for phrase in ["i don't know", "i don't have", "not sure", "cannot find", "no information"]):
            self._log("Initial response insufficient, performing web search")
            search_results = await self._run_in_executor(self.web_search_ollama, query)
            
            # Generate new response with web search results
            enhanced_prompt = f"""You are ArchieAI, an AI assistant for Arcadia University.

The user asked: {query}

Here is additional information from a web search:
{search_results}

Please provide a helpful answer based on this information.{history_context}"""
            
            return await self.generate_text_async(query, system_prompt=enhanced_prompt)
        
        return initial_response
    
    async def Archie_streaming(self, query: str, conversation_history: list = None) -> AsyncIterator[str]:
        """
        Streaming version of Archie that yields tokens as they are generated.
        This provides a better user experience by showing the AI "thinking" in real-time.
        
        Usage:
            async for token in ai.Archie_streaming("When is fall break?"):
                print(token, end='', flush=True)
        """
        with open("data/scrape_results.json", "r", encoding="utf-8") as f:
            results = json.load(f)
        
        # Build context with conversation history
        history_context = ""
        if conversation_history:
            history_context = "\n\nConversation History:\n"
            for msg in conversation_history[-5:]:  # Last 5 messages for context
                role = msg.get("role", "user")
                content = msg.get("content", "")
                history_context += f"{role.upper()}: {content}\n"
        
        # Try to determine if we need web search first
        # For streaming, we'll check keywords in the query itself
        needs_web_search = any(keyword in query.lower() for keyword in [
            "current", "latest", "recent", "today", "now", "news", "weather"
        ])
        
        if needs_web_search:
            self._log("Query suggests need for current info, performing web search")
            search_results = await self._run_in_executor(self.web_search_ollama, query)
            
            system_prompt = f"""You are ArchieAI, an AI assistant for Arcadia University.

Here is information from a web search:
{search_results}

Use the following university data as additional context:
{json.dumps(results, indent=2)}
{history_context}"""
        else:
            system_prompt = f"""You are ArchieAI, an AI assistant for Arcadia University. You are here to help students, faculty, and staff with any questions they may have about the university.

You are made by students for a final project. You must be factual and accurate based on the information provided. It is ok to say "I don't know" if you are unsure.

Respond based on your knowledge up to 2025.

Use the following content to better answer the query:
{json.dumps(results, indent=2)}
{history_context}"""

        async for token in self.generate_text_streaming(query, system_prompt=system_prompt):
            yield token
    

if __name__ == "__main__":
    # Simple test of the AiInterface
    ai = AiInterface(debug=True)
    test_query = "When is fall break?"
    response = asyncio.run(ai.Archie(test_query))
    print("Response to query:", test_query)
    print(response)

File: src/gem_interface.rs
// AI Interface using Ollama for local LLM inference with streaming support.

use ollama_rs::{
    generation::chat::{request::ChatMessageRequest, ChatMessage},
    Ollama,
};
use serde::{Deserialize, Serialize};
use std::env;
use tokio_stream::StreamExt;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Message {
    pub role: String,
    pub content: String,
}

impl Message {
    fn to_chat_message(&self) -> ChatMessage {
        match self.role.as_str() {
            "system" => ChatMessage::system(self.content.clone()),
            "assistant" => ChatMessage::assistant(self.content.clone()),
            _ => ChatMessage::user(self.content.clone()),
        }
    }
}

pub struct AiInterface {
    model: String,
    debug: bool,
    ollama_client: Ollama,
}

impl AiInterface {
    /// AI Interface using Ollama for local LLM inference with streaming support.
    ///
    /// Usage:
    ///   let ai = AiInterface::new(false, 3, 1.0, 15);
    ///   let result = ai.archie("When is fall break?").await;
    pub fn new(
        debug: bool,
        _scraper_max_retries: u32,
        _scraper_backoff_factor: f32,
        _scraper_timeout: u64,
    ) -> Self {
        // Load the variables from the .env file into the environment
        dotenv::dotenv().ok();

        // Retrieve the model name from environment (defaults to llama2 if not set)
        let model = env::var("MODEL").unwrap_or_else(|_| "llama2".to_string());

        // Create Ollama client - defaults to localhost:11434
        // This was recommended by some forms i found online to avoid connection issues
        let host = env::var("OLLAMA_HOST").unwrap_or_else(|_| "http://localhost".to_string());
        let port: u16 = env::var("OLLAMA_PORT")
            .ok()
            .and_then(|p| p.parse().ok())
            .unwrap_or(11434);

        let ollama_client = Ollama::new(host, port);

        AiInterface {
            model,
            debug,
            ollama_client,
        }
    }
    // A temporary logging function to designate debug messages
    fn log(&self, message: &str) {
        if self.debug {
            println!("[AiInterface DEBUG] {}", message);
        }
    }

    pub async fn async_web_search(
        &self,
        prompt: String,
        system_prompt: String,
    ) -> Result<Vec<String>, String> {
        self.log(&format!("Web search for: {}", prompt));

        // Check for OLLAMA_API_KEY or OLLAMA_TOKEN in environment
        let _ollama_api_key = env::var("OLLAMA_API_KEY")
            .or_else(|_| env::var("OLLAMA_TOKEN"))
            .ok();

        let model = env::var("OLLAMA_MODEL").unwrap_or_else(|_| self.model.clone());

        let mut messages = Vec::new();

        if !system_prompt.is_empty() {
            messages.push(ChatMessage::system(system_prompt));
        }

        messages.push(ChatMessage::user(prompt));

        let request = ChatMessageRequest::new(model, messages);

        match self.ollama_client.send_chat_messages_stream(request).await {
            Ok(mut stream) => {
                let mut results = Vec::new();
                while let Some(Ok(response)) = stream.next().await {
                    results.push(response.message.content);
                }
                Ok(results)
            }
            Err(e) => Err(format!("Failed to perform web search: {}", e)),
        }
    }

    pub async fn archie_streaming(
        &self,
        query: String,
        conversation_history: Option<Vec<Message>>,
    ) -> Result<Vec<String>, String> {
        self.log(&format!("Archie streaming query: {}", query));

        // Build context with conversation history
        let mut history_context = String::new();
        if let Some(history) = conversation_history {
            history_context.push_str("\n\nConversation History:\n");
            for msg in history.iter() {
                history_context.push_str(&format!(
                    "{}: {}\n",
                    msg.role.to_uppercase(),
                    msg.content
                ));
            }
        }

        let current_time = chrono::Utc::now().format("%Y-%m-%d %H:%M:%S").to_string();

        let system_prompt = format!(
            r#"You are ArchieAI, an AI assistant for Arcadia University. You are here to help students, faculty, and staff with any questions they may have about the university.

You are made by students for a final project. You must be factual and concise based on the information provided. All responses should be professional yet to the point.
Markdown IS NOT SUPPORTED OR RENDERED in the final output. DO NOT RESPOND WITH MARKDOWN FORMATTING OR HYPERLINKS so no [links](url) formatting or bolding. however you can provide full URLs.
You are not associated with Arcadia University officially as you are a student project.
History:
{}
The Time is {}"#,
            history_context, current_time
        );

        self.async_web_search(query, system_prompt).await
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_ai_interface_new() {
        let ai = AiInterface::new(false, 3, 1.0, 15);
        assert!(!ai.debug);
    }

    #[test]
    fn test_ai_interface_new_debug() {
        let ai = AiInterface::new(true, 3, 1.0, 15);
        assert!(ai.debug);
    }

    #[test]
    fn test_message_creation() {
        let msg = Message {
            role: "user".to_string(),
            content: "Hello".to_string(),
        };
        
        assert_eq!(msg.role, "user");
        assert_eq!(msg.content, "Hello");
    }

    #[test]
    fn test_message_to_chat_message_user() {
        let msg = Message {
            role: "user".to_string(),
            content: "Hello".to_string(),
        };
        
        let _chat_msg = msg.to_chat_message();
        // Just verify it doesn't panic
        assert!(true);
    }

    #[test]
    fn test_message_to_chat_message_assistant() {
        let msg = Message {
            role: "assistant".to_string(),
            content: "Hi there".to_string(),
        };
        
        let _chat_msg = msg.to_chat_message();
        assert!(true);
    }

    #[test]
    fn test_message_to_chat_message_system() {
        let msg = Message {
            role: "system".to_string(),
            content: "System prompt".to_string(),
        };
        
        let _chat_msg = msg.to_chat_message();
        assert!(true);
    }

    #[test]
    fn test_message_serialization() {
        let msg = Message {
            role: "user".to_string(),
            content: "Test message".to_string(),
        };
        
        let json = serde_json::to_string(&msg).expect("Failed to serialize");
        assert!(json.contains("user"));
        assert!(json.contains("Test message"));
    }

    #[test]
    fn test_message_deserialization() {
        let json = r#"{"role": "assistant", "content": "Response"}"#;
        let msg: Message = serde_json::from_str(json).expect("Failed to deserialize");
        
        assert_eq!(msg.role, "assistant");
        assert_eq!(msg.content, "Response");
    }

    #[test]
    fn test_message_clone() {
        let msg = Message {
            role: "user".to_string(),
            content: "Test".to_string(),
        };
        
        let cloned = msg.clone();
        assert_eq!(msg.role, cloned.role);
        assert_eq!(msg.content, cloned.content);
    }

    #[test]
    fn test_log_function_debug_mode() {
        let ai = AiInterface::new(true, 3, 1.0, 15);
        // This should not panic
        ai.log("Test debug message");
    }

    #[test]
    fn test_log_function_no_debug() {
        let ai = AiInterface::new(false, 3, 1.0, 15);
        // This should not panic or output anything
        ai.log("This should not be printed");
    }

    #[tokio::test]
    async fn test_archie_streaming_with_empty_history() {
        let ai = AiInterface::new(false, 3, 1.0, 15);
        // This test will fail without Ollama running, but we can test the structure
        // In a real environment, you'd mock the Ollama client
        let result = ai.archie_streaming(
            "Test query".to_string(),
            None,
        ).await;
        
        // We can't test the actual result without Ollama, but we can verify the function signature works
        assert!(result.is_ok() || result.is_err());
    }

    #[tokio::test]
    async fn test_archie_streaming_with_history() {
        let ai = AiInterface::new(false, 3, 1.0, 15);
        
        let history = vec![
            Message {
                role: "user".to_string(),
                content: "Previous question".to_string(),
            },
            Message {
                role: "assistant".to_string(),
                content: "Previous answer".to_string(),
            },
        ];
        
        let result = ai.archie_streaming(
            "Follow-up question".to_string(),
            Some(history),
        ).await;
        
        // We can't test the actual result without Ollama, but we can verify the function signature works
        assert!(result.is_ok() || result.is_err());
    }

    #[tokio::test]
    async fn test_async_web_search_structure() {
        let ai = AiInterface::new(false, 3, 1.0, 15);
        
        let result = ai.async_web_search(
            "Test query".to_string(),
            "Test system prompt".to_string(),
        ).await;
        
        // Can't test actual result without Ollama
        assert!(result.is_ok() || result.is_err());
    }

    #[test]
    fn test_multiple_message_conversions() {
        let messages = vec![
            Message { role: "system".to_string(), content: "System".to_string() },
            Message { role: "user".to_string(), content: "User".to_string() },
            Message { role: "assistant".to_string(), content: "Assistant".to_string() },
        ];
        
        for msg in messages {
            let _ = msg.to_chat_message();
        }
        
        assert!(true);
    }

    #[test]
    fn test_message_debug_format() {
        let msg = Message {
            role: "user".to_string(),
            content: "Test".to_string(),
        };
        
        let debug_string = format!("{:?}", msg);
        assert!(debug_string.contains("user"));
        assert!(debug_string.contains("Test"));
    }
}
